tools/clang/unittests/SPIRV/LibTestUtils.cpp

```diff
@@ -0,0 +1,228 @@
+      // Also push target environment to restArgs so that it gets passed to the
+      // compile command.
+      restArgs->push_back(tokens[i]);
+    } else {
+      restArgs->push_back(tokens[i]);
+    }
+  }
+
+  if (targetProfile->empty()) {
+    fprintf(stderr, "Error: Missing target profile argument (-T).\n");
+    return false;
+  }
+  // lib_6_* profile doesn't need an entryPoint
+  if (targetProfile->c_str()[0] != 'l' && entryPoint->empty()) {
+    fprintf(stderr, "Error: Missing entry point argument (-E).\n");
+    return false;
+  }
+  return true;
+}
+
+void convertIDxcBlobToUint32(const CComPtr<IDxcBlob> &blob,
+                             std::vector<uint32_t> *binaryWords) {
+  size_t num32BitWords = (blob->GetBufferSize() + 3) / 4;
+  std::string binaryStr((char *)blob->GetBufferPointer(),
+                        blob->GetBufferSize());
+  binaryStr.resize(num32BitWords * 4, 0);
+  binaryWords->resize(num32BitWords, 0);
+  memcpy(binaryWords->data(), binaryStr.data(), binaryStr.size());
+}
+
+bool compileCodeWithSpirvGeneration(const llvm::StringRef code,
+                                    const llvm::StringRef entryPoint,
+                                    const llvm::StringRef targetProfile,
+                                    const std::vector<std::string> &restArgs,
+                                    std::vector<uint32_t> *generatedBinary,
+                                    std::string *errorMessages) {
+  std::wstring srcFile(L"");
+  std::wstring entry(entryPoint.begin(), entryPoint.end());
+  std::wstring profile(targetProfile.begin(), targetProfile.end());
+
+  std::vector<std::wstring> rest;
+  for (const auto &arg : restArgs)
+    rest.emplace_back(arg.begin(), arg.end());
+
+  bool success = true;
+
+  try {
+    dxc::DxcDllSupport dllSupport;
+    IFT(dllSupport.Initialize());
+
+    if (hlsl::options::initHlslOptTable())
+      throw std::bad_alloc();
+
+    CComPtr<IDxcLibrary> pLibrary;
+    CComPtr<IDxcCompiler> pCompiler;
+    CComPtr<IDxcOperationResult> pResult;
+    CComPtr<IDxcBlobEncoding> pErrorBuffer;
+    CComPtr<IDxcBlob> pCompiledBlob;
+    CComPtr<IDxcIncludeHandler> pIncludeHandler;
+    HRESULT resultStatus;
+
+    bool requires_opt = false;
+    for (const auto &arg : rest)
+      if (arg == L"-O3" || arg == L"-O0" || arg.substr(0, 8) == L"-Oconfig")
+        requires_opt = true;
+
+    std::vector<LPCWSTR> flags;
+    // lib_6_* profile doesn't need an entryPoint
+    if (profile.c_str()[0] != 'l') {
+      flags.push_back(L"-E");
+      flags.push_back(entry.c_str());
+    }
+    flags.push_back(L"-T");
+    flags.push_back(profile.c_str());
+    flags.push_back(L"-spirv");
+    // Disable legalization and optimization for testing, unless the caller
+    // wants to run a specific optimization recipe (with -Oconfig).
+    if (!requires_opt)
+      flags.push_back(L"-fcgl");
+    // Disable validation. We'll run it manually.
+    flags.push_back(L"-Vd");
+    for (const auto &arg : rest)
```

Comment:
Shall the logic be more "forward the given arguments as-is"?
(and adding `-Vd` to be able to manually validate later on)

Reply:
For test, you want to caller to deal with validation so that it can parse the errors if they are expected. Again, this is bringing back the old code with as few changes as possible. If we want to change this it can be done later.